src/status_emitter/json.rs

```diff
@@ -13,54 +13,68 @@ use bstr::ByteSlice;
 
 // MAINTENANCE REGION START
 
-// When integrating with a new libtest version, update all emit_xxx functions.
+// When integrating with a new libtest version, update all xxx_event functions.
 
-fn emit_suite_end(failed: usize, filtered_out: usize, ignored: usize, passed: usize, status: &str) {
+fn suite_end_event(
+    failed: usize,
+    filtered_out: usize,
+    ignored: usize,
+    passed: usize,
+    status: &str,
+) -> String {
     // Adapted from test::formatters::json::write_run_finish().
-    println!(
+    format!(
         r#"{{ "type": "suite", "event": "{status}", "passed": {passed}, "failed": {failed}, "ignored": {ignored}, "measured": 0, "filtered_out": {filtered_out} }}"#
-    );
+    )
 }
 
-fn emit_suite_start() {
+fn suite_start_event() -> String {
     // Adapted from test::formatters::json::write_run_start().
-    println!(r#"{{ "type": "suite", "event": "started" }}"#);
+    String::from(r#"{ "type": "suite", "event": "started" }"#)
 }
 
-fn emit_test_end(name: &String, revision: &String, path: &Path, status: &str, diags: &str) {
-    let displayed_path = path.display();
-    let stdout = if diags.is_empty() {
+fn test_end_event(name: &str, revision: &str, path: &Path, status: &str, diags: &str) -> String {
+    let escaped_name = escape_backslashes(name);
+    let escaped_revision = escape_backslashes(revision);
+    let escaped_path = escape_backslashes(&path.display().to_string());
+    let escaped_stdout = if diags.is_empty() {
         String::new()
     } else {
-        let triaged_diags = serde_json::to_string(diags).unwrap();
-        format!(r#", "stdout": {triaged_diags}"#)
+        let escaped_diags = serde_json::to_string(&escape_backslashes(diags)).unwrap();
+        format!(r#", "stdout": {}"#, escaped_diags)
     };
 
     // Adapted from test::formatters::json::write_event().
-    println!(
-        r#"{{ "type": "test", "event": "{status}", "name": "{name} ({revision}) - {displayed_path}"{stdout} }}"#
-    );
+    format!(
+        r#"{{ "type": "test", "event": "{status}", "name": "{escaped_name} ({escaped_revision}) - {escaped_path}"{escaped_stdout} }}"#
+    )
 }
 
-fn emit_test_start(name: &String, revision: &String, path: &Path) {
-    let displayed_path = path.display();
+fn test_start_event(name: &str, revision: &str, path: &Path) -> String {
+    let escaped_name = escape_backslashes(name);
+    let escaped_revision = escape_backslashes(revision);
+    let escaped_path = escape_backslashes(&path.display().to_string());
 
     // Adapted from test::formatters::json::write_test_start().
-    println!(
-        r#"{{ "type": "test", "event": "started", "name": "{name} ({revision}) - {displayed_path}" }}"#
-    );
+    format!(
+        r#"{{ "type": "test", "event": "started", "name": "{escaped_name} ({escaped_revision}) - {escaped_path}" }}"#
+    )
 }
 
 // MAINTENANCE REGION END
 
+fn escape_backslashes(s: &str) -> String {
```

Comment:
Is that correct? Don't you need to escape everything?

Reply:
Done. The code actually ended up looking way better!